File: AMF5/src/test/scala/scalaPlatform/ParsingTest.scala
package scalaPlatform

import amf.apicontract.client.scala.{
  APIConfiguration,
  AsyncAPIConfiguration,
  OASConfiguration,
  RAMLConfiguration,
  WebAPIConfiguration
}
import amf.core.client.scala.model.document.Document
import amf.core.internal.remote.Spec
import org.junit.runner.RunWith
import org.scalatest.flatspec.AsyncFlatSpec
import org.scalatest.matchers.must.Matchers.convertToAnyMustWrapper
import org.scalatest.matchers.should
import org.scalatestplus.junit.JUnitRunner
import utils.BaseUnitUtils.{
  getFirstOperationFromLastEndpoint,
  getFirstRequest,
  getFirstServer,
  getLastEndpoint,
  getLastServer,
  getMessageComponent
}

@RunWith(classOf[JUnitRunner])
class ParsingTest extends AsyncFlatSpec with should.Matchers {

  "AMF client" should "parse an OAS 2.0 API" in {
    val client = OASConfiguration.OAS20().baseUnitClient()
    client.parse("file://src/test/resources/examples/banking-api.json") map { result =>
      result.baseUnit mustBe a[Document]
      result.conforms shouldBe true
    }
  }

  it should "parse an OAS 2.0 API from a string" in {
    val client = OASConfiguration.OAS20().baseUnitClient()
    val api =
      """{
        |    "swagger": "2.0",
        |    "info": {
        |        "title": "ACME Banking HTTP API",
        |        "version": "1.0"
        |    },
        |    "host": "acme-banking.com",
        |    "paths": {}
        |}""".stripMargin
    client.parseContent(api) map { result =>
      result.baseUnit mustBe a[Document]
      result.conforms shouldBe true
    }
  }

  it should "parse an OAS 3.0 API" in {
    val client = OASConfiguration.OAS30().baseUnitClient()
    client.parse(
      "file://src/test/resources/examples/banking-api-oas30.json"
    ) map { result =>
      result.baseUnit mustBe a[Document]
      result.conforms shouldBe true
    }
  }

  it should "parse an OAS 3.0 API from a string" in {
    val client = OASConfiguration.OAS30().baseUnitClient()
    val api =
      """{
        |  "openapi": "3.0.0",
        |  "info": {
        |    "title": "Basic content",
        |    "version": "0.1"
        |  },
        |  "paths": {}
        |}""".stripMargin
    client.parseContent(api) map { result =>
      result.baseUnit mustBe a[Document]
      result.conforms shouldBe true
    }
  }

  it should "parse an RAML 1.0 API" in {
    val client = RAMLConfiguration.RAML10().baseUnitClient()
    client.parse("file://src/test/resources/examples/banking-api.raml") map { result =>
      result.baseUnit mustBe a[Document]
      result.conforms shouldBe true
    }
  }

  it should "parse an RAML 1.0 API from a string" in {
    val client = RAMLConfiguration.RAML10().baseUnitClient()
    val api =
      """#%RAML 1.0
        |
        |title: ACME Banking HTTP API
        |version: 1.0""".stripMargin
    client.parseContent(api) map { result =>
      result.baseUnit mustBe a[Document]
      result.conforms shouldBe true
    }
  }

  it should "parse an RAML 0.8 API" in {
    val client = RAMLConfiguration.RAML08().baseUnitClient()
    client.parse(
      "file://src/test/resources/examples/banking-api-08.raml"
    ) map { result =>
      result.baseUnit mustBe a[Document]
      result.conforms shouldBe true
    }
  }

  it should "parse an RAML 0.8 API from a string" in {
    val client = RAMLConfiguration.RAML08().baseUnitClient()
    val api =
      """#%RAML 0.8
        |
        |title: ACME Banking HTTP API
        |version: 1.0""".stripMargin
    client.parseContent(api) map { result =>
      result.baseUnit mustBe a[Document]
      result.conforms shouldBe true
    }
  }

  it should "parse an unknown WebAPI" in {
    val client = WebAPIConfiguration.WebAPI().baseUnitClient()
    client.parse("file://src/test/resources/examples/banking-api.raml") map { result =>
      result.baseUnit mustBe a[Document]
      result.conforms shouldBe true
      result.sourceSpec.isRaml shouldBe true
      result.sourceSpec.id mustBe Spec.RAML10.id
    }
  }

  it should "parse an unknown WebAPI from a string" in {
    val client = WebAPIConfiguration.WebAPI().baseUnitClient()
    val api =
      """{
        |  "openapi": "3.0.0",
        |  "info": {
        |    "title": "Basic content",
        |    "version": "0.1"
        |  },
        |  "paths": {}
        |}""".stripMargin
    client.parseContent(api) map { result =>
      result.baseUnit mustBe a[Document]
      result.conforms shouldBe true
      result.sourceSpec.isOas shouldBe true
      result.sourceSpec.id mustBe Spec.OAS30.id
    }
  }

  it should "parse an unknown API" in {
    val client = APIConfiguration.API().baseUnitClient()
    client.parse("file://src/test/resources/examples/async.yaml") map { result =>
      result.baseUnit mustBe a[Document]
      result.conforms shouldBe true
      result.sourceSpec.isAsync shouldBe true
      result.sourceSpec.id mustBe Spec.ASYNC20.id
    }
  }

  it should "parse an unknown API from a string" in {
    val client = APIConfiguration.API().baseUnitClient()
    val api =
      """asyncapi: "2.0.0"
        |info:
        |  title: "Something"
        |  version: "1.0"
        |channels: {}
        |""".stripMargin
    client.parseContent(api) map { result =>
      result.baseUnit mustBe a[Document]
      result.conforms shouldBe true
      result.sourceSpec.isAsync shouldBe true
      result.sourceSpec.id mustBe Spec.ASYNC20.id
    }
  }

  it should "parse an AsyncAPI 2.1 API" in {
    val client = AsyncAPIConfiguration.Async20().baseUnitClient()
    client.parse("file://src/test/resources/examples/asyncApi-2.1-all.yaml") map { result =>
      val baseUnit = result.baseUnit.asInstanceOf[Document]
      // This version supports Mercure and IBMMQ bindings

      // Server Binding
      val server = getFirstServer(baseUnit)
      val ibmmqServerBinding = server.bindings.bindings.head

      // Channel Binding
      val someOtherChannel = getLastEndpoint(baseUnit)
      val ibmmqChannelBinding = someOtherChannel.bindings.bindings.head

      // Message Binding
      val messageRequest = getFirstRequest(baseUnit)
      val ibmmqMessageBinding = messageRequest.bindings.bindings.head

      result.baseUnit mustBe a[Document]
      result.conforms shouldBe true
      result.sourceSpec.isAsync shouldBe true
      result.sourceSpec.id mustBe Spec.ASYNC21.id
      ibmmqChannelBinding.componentId shouldBe "/ibmmq-channel"
      ibmmqServerBinding.componentId shouldBe "/ibmmq-server"
      ibmmqMessageBinding.componentId shouldBe "/ibmmq-message"
    }
  }

  it should "parse an AsyncAPI 2.2 API" in {
    val client = AsyncAPIConfiguration.Async20().baseUnitClient()
    client.parse("file://src/test/resources/examples/asyncApi-2.2-all.yaml") map { result =>
      val baseUnit = result.baseUnit.asInstanceOf[Document]
      // This version adds AnypointMQ bindings

      // Channel Binding
      val anotherChannel = getLastEndpoint(baseUnit)
      val anypointMQChannelBinding = anotherChannel.bindings.bindings.head

      // Message Binding
      val messageRequest = getFirstRequest(baseUnit)
      val anypointMQMessageBinding = messageRequest.bindings.bindings.head

      result.baseUnit mustBe a[Document]
      result.conforms shouldBe true
      result.sourceSpec.isAsync shouldBe true
      result.sourceSpec.id mustBe Spec.ASYNC22.id
      anypointMQChannelBinding.componentId shouldBe "/anypointmq-channel"
      anypointMQMessageBinding.componentId shouldBe "/anypointmq-message"
    }
  }

  it should "parse an AsyncAPI 2.3 API" in {
    val client = AsyncAPIConfiguration.Async20().baseUnitClient()
    client.parse("file://src/test/resources/examples/asyncApi-2.3-all.yaml") map { result =>
      val baseUnit = result.baseUnit.asInstanceOf[Document]
      // This version adds Solace bindings

      // Operation Binding
      val publishOperation = getFirstOperationFromLastEndpoint(baseUnit)
      val solaceOperationBinding = publishOperation.bindings.bindings.head

      result.baseUnit mustBe a[Document]
      result.conforms shouldBe true
      result.sourceSpec.isAsync shouldBe true
      result.sourceSpec.id mustBe Spec.ASYNC23.id
      // Asume 0.3.0 version
      solaceOperationBinding.componentId shouldBe "/solace-operation-030"
    }
  }

  it should "parse an AsyncAPI 2.4 API" in {
    val client = AsyncAPIConfiguration.Async20().baseUnitClient()
    client.parse("file://src/test/resources/examples/asyncApi-2.4-all.yaml") map { result =>
      //no new bindings in this version
      result.baseUnit mustBe a[Document]
      result.conforms shouldBe true
      result.sourceSpec.isAsync shouldBe true
      result.sourceSpec.id mustBe Spec.ASYNC24.id
    }
  }

  it should "parse an AsyncAPI 2.5 API" in {
    val client = AsyncAPIConfiguration.Async20().baseUnitClient()
    client.parse("file://src/test/resources/examples/asyncApi-2.5-all.yaml") map { result =>
      val baseUnit = result.baseUnit.asInstanceOf[Document]
      // This version adds GooglePubSub binding.

      // Channel Binding
      val topicProtoSchema = getLastEndpoint(baseUnit)
      val googlePubSubChannelBinding = topicProtoSchema.bindings.bindings.head

      // Message Binding
      val messageComponent = getMessageComponent(baseUnit)
      val googlePubSubMessageBinding = messageComponent.bindings.bindings.head

      result.baseUnit mustBe a[Document]
      result.conforms shouldBe true
      result.sourceSpec.isAsync shouldBe true
      result.sourceSpec.id mustBe Spec.ASYNC25.id
      // By default we assume that the version is 0.1.0
      googlePubSubChannelBinding.componentId shouldBe "/googlepubsub-channel-010"
      googlePubSubMessageBinding.componentId shouldBe "/googlepubsub-message-010"
    }
  }

  it should "parse an AsyncAPI 2.6 API" in {
    val client = AsyncAPIConfiguration.Async20().baseUnitClient()
    client.parse("file://src/test/resources/examples/asyncApi-2.6-all.yaml") map { result =>
      val baseUnit = result.baseUnit.asInstanceOf[Document]
      // This version adds Pulsar bindings.

      // Server Binding
      val theNameServer = getLastServer(baseUnit)
      val pulsarServerBinding = theNameServer.bindings.bindings.head

      // Channel Binding
      val sixthChannel = getLastEndpoint(baseUnit)
      val pulsarChannelBinding = sixthChannel.bindings.bindings.head

      result.baseUnit mustBe a[Document]
      result.conforms shouldBe true
      result.sourceSpec.isAsync shouldBe true
      result.sourceSpec.id mustBe Spec.ASYNC26.id
      pulsarServerBinding.componentId shouldBe "/pulsar-server"
    // Uncomment this line when bug is fixed
//      pulsarChannelBinding.componentId shouldBe "/pulsar-channel"
    }
  }
}

File: AMF5/src/test/java/javaPlatform/ParsingTest.java
package javaPlatform;

import amf.apicontract.client.platform.*;
import amf.apicontract.client.platform.model.domain.*;
import amf.apicontract.client.platform.model.domain.api.AsyncApi;
import amf.apicontract.client.platform.model.domain.bindings.ChannelBinding;
import amf.apicontract.client.platform.model.domain.bindings.MessageBinding;
import amf.apicontract.client.platform.model.domain.bindings.OperationBinding;
import amf.apicontract.client.platform.model.domain.bindings.ServerBinding;
import amf.core.client.platform.AMFParseResult;
import amf.core.client.platform.model.document.BaseUnit;
import amf.core.client.platform.model.document.Document;
import amf.core.client.platform.model.domain.DomainElement;
import amf.core.internal.remote.Spec;
import org.junit.Test;
import utils.BaseUnitUtilsJava;

import java.util.List;
import java.util.concurrent.ExecutionException;

import static org.junit.Assert.*;

public class ParsingTest {

    @Test
    public void parseOas20() throws ExecutionException, InterruptedException {
        final AMFBaseUnitClient client = OASConfiguration.OAS20().baseUnitClient();

        // A BaseUnit is the return type of any parsing
        // The actual object can be many different things, depending on the content of the source file
        // https://github.com/aml-org/amf/blob/develop/documentation/model.md#baseunit
        final AMFParseResult parseResult = client.parse("file://src/test/resources/examples/banking-api.json").get();
        final BaseUnit model = parseResult.baseUnit();

        assertNotNull(model);
        assertTrue(parseResult.conforms());

        // DomainElement is the base class for any element describing a domain model
        final DomainElement webApi = ((Document) model).encodes();
        assertNotNull(webApi);
    }

    @Test
    public void parseOas20String() throws ExecutionException, InterruptedException {
        final AMFBaseUnitClient client = OASConfiguration.OAS20().baseUnitClient();

        final String api =
                "{\n" +
                        "  \"swagger\": \"2.0\",\n" +
                        "  \"info\": {\n" +
                        "    \"title\": \"ACME Banking HTTP API\",\n" +
                        "    \"version\": \"1.0\"\n" +
                        "  },\n" +
                        "  \"host\": \"acme-banking.com\",\n" +
                        "  \"paths\": {}\n" +
                        "}";

        final AMFParseResult parseResult = client.parseContent(api).get();
        final BaseUnit model = parseResult.baseUnit();

        assertNotNull(model);
        assertTrue(parseResult.conforms());
        assertTrue(model.raw().isPresent());
        assertEquals(model.raw().get(), api);

        final DomainElement webApi = ((Document) model).encodes();
        assertNotNull(webApi);
    }

    @Test
    public void parseOas30() throws ExecutionException, InterruptedException {
        final AMFBaseUnitClient client = OASConfiguration.OAS30().baseUnitClient();

        final AMFParseResult parseResult = client.parse("file://src/test/resources/examples/banking-api-oas30.json").get();
        final BaseUnit model = parseResult.baseUnit();

        assertNotNull(model);
        assertTrue(parseResult.conforms());
        final DomainElement webApi = ((Document) model).encodes();
        assertNotNull(webApi);
    }

    @Test
    public void parseUnknownWebAPI() throws ExecutionException, InterruptedException {
        final AMFBaseUnitClient client = WebAPIConfiguration.WebAPI().baseUnitClient();

        final AMFParseResult parseResult = client.parse("file://src/test/resources/examples/banking-api-oas30.json").get();
        final BaseUnit model = parseResult.baseUnit();
        final Spec spec = parseResult.sourceSpec();

        assertNotNull(model);
        assertTrue(parseResult.conforms());
        final DomainElement webApi = ((Document) model).encodes();
        assertNotNull(webApi);
        assertTrue(spec.isOas());
        assertEquals(spec.id(), Spec.OAS30().id());
    }

    @Test
    public void parseUnknownAPI() throws ExecutionException, InterruptedException {
        final AMFBaseUnitClient client = APIConfiguration.API().baseUnitClient();

        final AMFParseResult parseResult = client.parse("file://src/test/resources/examples/async.yaml").get();
        final BaseUnit model = parseResult.baseUnit();
        final Spec spec = parseResult.sourceSpec();

        assertNotNull(model);
        assertTrue(parseResult.conforms());
        final DomainElement asyncApi = ((Document) model).encodes();
        assertNotNull(asyncApi);
        assertTrue(spec.isAsync());
        assertEquals(spec.id(), Spec.ASYNC20().id());
    }

    @Test
    public void parseOas30String() throws ExecutionException, InterruptedException {
        final AMFBaseUnitClient client = OASConfiguration.OAS30().baseUnitClient();

        final String api = "{\n" +
                "  \"openapi\": \"3.0.0\",\n" +
                "  \"info\": {\n" +
                "    \"title\": \"Basic content\",\n" +
                "    \"version\": \"0.1\"\n" +
                "  },\n" +
                "  \"paths\": {}\n" +
                "}";

        final AMFParseResult parseResult = client.parseContent(api).get();
        final BaseUnit model = parseResult.baseUnit();

        assertNotNull(model);
        assertTrue(parseResult.conforms());
        assertTrue(model.raw().isPresent());
        assertEquals(model.raw().get(), api);

        final DomainElement webApi = ((Document) model).encodes();
        assertNotNull(webApi);
    }

    @Test
    public void parseRaml10() throws ExecutionException, InterruptedException {
        final AMFBaseUnitClient client = RAMLConfiguration.RAML10().baseUnitClient();

        final AMFParseResult parseResult = client.parse("file://src/test/resources/examples/banking-api.raml").get();
        final BaseUnit model = parseResult.baseUnit();

        assertNotNull(model);
        assertTrue(parseResult.conforms());

        final DomainElement webApi = ((Document) model).encodes();
        assertNotNull(webApi);
    }

    @Test
    public void parseRaml10String() throws ExecutionException, InterruptedException {
        final AMFBaseUnitClient client = RAMLConfiguration.RAML10().baseUnitClient();

        final String api =
                "#%RAML 1.0\n" +
                        "\n" +
                        "title: ACME Banking HTTP API\n" +
                        "version: 1.0";

        final AMFParseResult parseResult = client.parseContent(api).get();
        final BaseUnit model = parseResult.baseUnit();

        assertNotNull(model);
        assertTrue(parseResult.conforms());
        assertTrue(model.raw().isPresent());
        assertEquals(model.raw().get(), api);

        final DomainElement webApi = ((Document) model).encodes();
        assertNotNull(webApi);
    }

    @Test
    public void parseRaml08() throws ExecutionException, InterruptedException {
        final AMFBaseUnitClient client = RAMLConfiguration.RAML08().baseUnitClient();

        final AMFParseResult parseResult = client.parse("file://src/test/resources/examples/banking-api-08.raml").get();
        final BaseUnit model = parseResult.baseUnit();

        assertNotNull(model);
        assertTrue(parseResult.conforms());

        final DomainElement webApi = ((Document) model).encodes();
        assertNotNull(webApi);
    }

    @Test
    public void parseRaml08String() throws ExecutionException, InterruptedException {
        final AMFBaseUnitClient client = RAMLConfiguration.RAML08().baseUnitClient();

        final String api =
                "#%RAML 0.8\n" +
                        "\n" +
                        "title: ACME Banking HTTP API\n" +
                        "version: 1.0";

        final AMFParseResult parseResult = client.parseContent(api).get();
        final BaseUnit model = parseResult.baseUnit();

        assertNotNull(model);
        assertTrue(parseResult.conforms());
        assertTrue(model.raw().isPresent());
        assertEquals(model.raw().get(), api);

        final DomainElement webApi = ((Document) model).encodes();
        assertNotNull(webApi);
    }

    @Test
    public void parseUnknownWebAPIString() throws ExecutionException, InterruptedException {
        final AMFBaseUnitClient client = WebAPIConfiguration.WebAPI().baseUnitClient();

        final String api =
                "#%RAML 0.8\n" +
                        "\n" +
                        "title: ACME Banking HTTP API\n" +
                        "version: 1.0";

        final AMFParseResult parseResult = client.parseContent(api).get();
        final BaseUnit model = parseResult.baseUnit();
        final Spec spec = parseResult.sourceSpec();

        assertNotNull(model);
        assertTrue(parseResult.conforms());
        assertTrue(model.raw().isPresent());
        assertEquals(model.raw().get(), api);
        assertTrue(spec.isRaml());
        assertEquals(spec.id(), Spec.RAML08().id());

        final DomainElement webApi = ((Document) model).encodes();
        assertNotNull(webApi);
    }

    @Test
    public void parseUnknownAPIString() throws ExecutionException, InterruptedException {
        final AMFBaseUnitClient client = APIConfiguration.API().baseUnitClient();

        final String api =
                "asyncapi: \"2.0.0\"\n" +
                        "info:\n" +
                        "  title: \"Something\"\n" +
                        "  version: \"1.0\"\n" +
                        "channels: {}";

        final AMFParseResult parseResult = client.parseContent(api).get();
        final BaseUnit model = parseResult.baseUnit();
        final Spec spec = parseResult.sourceSpec();

        assertNotNull(model);
        assertTrue(parseResult.conforms());
        assertTrue(model.raw().isPresent());
        assertEquals(model.raw().get(), api);
        assertTrue(spec.isAsync());
        assertEquals(spec.id(), Spec.ASYNC20().id());

        final DomainElement asyncApi = ((Document) model).encodes();
        assertNotNull(asyncApi);
    }

    @Test
    public void parseAsyncApi21() throws ExecutionException, InterruptedException {
        final AMFBaseUnitClient client = AsyncAPIConfiguration.Async20().baseUnitClient();

        final AMFParseResult parseResult = client.parse("file://src/test/resources/examples/asyncApi-2.1-all.yaml").get();
        final BaseUnit baseUnit = parseResult.baseUnit();
        final Spec spec = parseResult.sourceSpec();
        final AsyncApi asyncApi = (AsyncApi) BaseUnitUtilsJava.getApi(baseUnit, true);
        // This version supports Mercure and IBMMQ bindings

        // Server Binding
        final Server server = BaseUnitUtilsJava.getFirstServer(baseUnit, true);
        final ServerBinding serverBinding = server.bindings().bindings().get(0);

        // Channel Binding
        final EndPoint someOtherChannel = BaseUnitUtilsJava.getLastEndPoint(baseUnit, true);
        final ChannelBinding channelBinding = someOtherChannel.bindings().bindings().get(0);

        // Message Binding
        final Request messageRequest = BaseUnitUtilsJava.getFirstRequest(baseUnit, true);
        final MessageBinding messageBinding = messageRequest.bindings().bindings().get(0);


        assertNotNull(baseUnit);
        assertTrue(parseResult.conforms());
        assertNotNull(asyncApi);
        assertTrue(spec.isAsync());
        assertEquals(spec.id(), Spec.ASYNC21().id());
        assertTrue(serverBinding.id().contains("/ibmmq-server"));
        assertTrue(channelBinding.id().contains("/ibmmq-channel"));
        assertTrue(messageBinding.id().contains("/ibmmq-message"));
    }

    @Test
    public void parseAsyncApi22() throws ExecutionException, InterruptedException {
        final AMFBaseUnitClient client = AsyncAPIConfiguration.Async20().baseUnitClient();

        final AMFParseResult parseResult = client.parse("file://src/test/resources/examples/asyncApi-2.2-all.yaml").get();
        final BaseUnit baseUnit = parseResult.baseUnit();
        final Spec spec = parseResult.sourceSpec();
        final AsyncApi asyncApi = (AsyncApi) BaseUnitUtilsJava.getApi(baseUnit, true);
        // This version adds AnypointMQ bindings


        // Channel Binding
        final List<EndPoint> endpoints = asyncApi.endPoints();
        final EndPoint anotherChannel = BaseUnitUtilsJava.getLastEndPoint(baseUnit, true);
        final ChannelBinding channelBinding = anotherChannel.bindings().bindings().get(0);

        // Message Binding
        final Request messageRequest = BaseUnitUtilsJava.getFirstRequest(baseUnit, true);
        final MessageBinding messageBinding = messageRequest.bindings().bindings().get(0);

        assertNotNull(baseUnit);
        assertTrue(parseResult.conforms());
        assertNotNull(asyncApi);
        assertTrue(spec.isAsync());
        assertEquals(spec.id(), Spec.ASYNC22().id());
        assertTrue(channelBinding.id().contains("/anypointmq-channel"));
        assertTrue(messageBinding.id().contains("/anypointmq-message"));

    }

    @Test
    public void parseAsyncApi23() throws ExecutionException, InterruptedException {
        final AMFBaseUnitClient client = AsyncAPIConfiguration.Async20().baseUnitClient();

        final AMFParseResult parseResult = client.parse("file://src/test/resources/examples/asyncApi-2.3-all.yaml").get();
        final BaseUnit baseUnit = parseResult.baseUnit();
        final Spec spec = parseResult.sourceSpec();
        final AsyncApi asyncApi = (AsyncApi) BaseUnitUtilsJava.getApi(baseUnit, true);
        // This version adds Solace bindings

        // Operation Binding
        final Operation publishOperation = BaseUnitUtilsJava.getFirstOperationFromLastEndpoint(baseUnit, true);
        final OperationBinding operationBinding = publishOperation.bindings().bindings().get(0);


        assertNotNull(baseUnit);
        assertTrue(parseResult.conforms());
        assertNotNull(asyncApi);
        assertTrue(spec.isAsync());
        assertEquals(spec.id(), Spec.ASYNC23().id());
        // Asume 0.3.0 version
        assertTrue(operationBinding.id().contains("/solace-operation-030"));
    }

    @Test
    public void parseAsyncApi24() throws ExecutionException, InterruptedException {
        final AMFBaseUnitClient client = AsyncAPIConfiguration.Async20().baseUnitClient();

        final AMFParseResult parseResult = client.parse("file://src/test/resources/examples/asyncApi-2.4-all.yaml").get();
        final BaseUnit baseUnit = parseResult.baseUnit();
        final Spec spec = parseResult.sourceSpec();
        // No new bindings in this version

        assertNotNull(baseUnit);
        assertTrue(parseResult.conforms());
        final DomainElement asyncApi = ((Document) baseUnit).encodes();
        assertNotNull(asyncApi);
        assertTrue(spec.isAsync());
        assertEquals(spec.id(), Spec.ASYNC24().id());
    }

    @Test
    public void parseAsyncApi25() throws ExecutionException, InterruptedException {
        final AMFBaseUnitClient client = AsyncAPIConfiguration.Async20().baseUnitClient();

        final AMFParseResult parseResult = client.parse("file://src/test/resources/examples/asyncApi-2.5-all.yaml").get();
        final BaseUnit baseUnit = parseResult.baseUnit();
        final Spec spec = parseResult.sourceSpec();
        final AsyncApi asyncApi = (AsyncApi) BaseUnitUtilsJava.getApi(baseUnit, true);
        // This version adds GooglePubSub binding.

        // Channel Binding
        final EndPoint topicProtoSchema = BaseUnitUtilsJava.getLastEndPoint(baseUnit, true);
        final ChannelBinding channelBinding = topicProtoSchema.bindings().bindings().get(0);

        // Message Binding
        final Message messageComponent = BaseUnitUtilsJava.getMessageComponent(baseUnit);
        final MessageBinding messageBinding = messageComponent.bindings().bindings().get(0);

        assertNotNull(baseUnit);
        assertTrue(parseResult.conforms());
        assertNotNull(asyncApi);
        assertTrue(spec.isAsync());
        assertEquals(spec.id(), Spec.ASYNC25().id());
        // By default, the version will be 0.1.0
        assertTrue(channelBinding.id().contains("/googlepubsub-channel-010"));
        assertTrue(messageBinding.id().contains("/googlepubsub-message-010"));
    }

    @Test
    public void parseAsyncApi26() throws ExecutionException, InterruptedException {
        final AMFBaseUnitClient client = AsyncAPIConfiguration.Async20().baseUnitClient();

        final AMFParseResult parseResult = client.parse("file://src/test/resources/examples/asyncApi-2.6-all.yaml").get();
        final BaseUnit baseUnit = parseResult.baseUnit();
        final Spec spec = parseResult.sourceSpec();
        final AsyncApi asyncApi = (AsyncApi) BaseUnitUtilsJava.getApi(baseUnit, true);
        // This version adds Pulsar bindings.

        // Server Binding
        final Server server = BaseUnitUtilsJava.getLastServer(baseUnit, true);
        final ServerBinding serverBinding = server.bindings().bindings().get(0);

        // Channel Binding
        final EndPoint sixthChannel = BaseUnitUtilsJava.getLastEndPoint(baseUnit, true);
        final ChannelBinding channelBinding = sixthChannel.bindings().bindings().get(0);

        assertNotNull(baseUnit);
        assertTrue(parseResult.conforms());
        assertNotNull(asyncApi);
        assertTrue(spec.isAsync());
        assertEquals(spec.id(), Spec.ASYNC26().id());
        assertTrue(serverBinding.id().contains("/pulsar-server"));
        // Uncomment this line when bug is fixed
//        assertTrue(channelBinding.id().contains("/pulsar-channel"));
    }
}
